diff --git a/include/vfs/Core/vfs_debug.h b/include/vfs/Core/vfs_debug.h
index 7cd285ada..32a7218d3 100644
--- a/include/vfs/Core/vfs_debug.h
+++ b/include/vfs/Core/vfs_debug.h
@@ -30,7 +30,7 @@
 #include <vfs/Core/vfs_path.h>
 #include <list>
 
-#ifdef _MSC_VER
+#if defined(_MSC_VER) && !defined(__clang__)
 class VFS_API std::exception;
 #endif
 
diff --git a/include/vfs/Core/vfs_string.h b/include/vfs/Core/vfs_string.h
index d3319a9d7..bcf5d293b 100644
--- a/include/vfs/Core/vfs_string.h
+++ b/include/vfs/Core/vfs_string.h
@@ -28,6 +28,7 @@
 #include <vfs/vfs_config.h>
 #include <string>
 #include <sstream>
+#include <algorithm>
 
 namespace vfs
 {
@@ -187,6 +188,12 @@ public:
 		_strstr.str(bs._strstr.str());
 	}
 
+	BuildString& operator=(const BuildString& bs)
+	{
+		_strstr.str(bs._strstr.str());
+		return *this;
+	}
+
 	template<typename T>
 	BuildString(T const& t)
 	{
diff --git a/include/vfs/Tools/vfs_hp_timer.h b/include/vfs/Tools/vfs_hp_timer.h
index dd198cf51..6272ebbd6 100644
--- a/include/vfs/Tools/vfs_hp_timer.h
+++ b/include/vfs/Tools/vfs_hp_timer.h
@@ -27,7 +27,7 @@
 #include <vfs/vfs_config.h>
 
 #ifdef WIN32
-	#include <Windows.h>
+	#include <windows.h>
 #elif __linux__
   #include <sys/time.h>
 #endif
diff --git a/include/vfs/Tools/vfs_parser_tools.h b/include/vfs/Tools/vfs_parser_tools.h
index 38bf53b2f..fd2697e7d 100644
--- a/include/vfs/Tools/vfs_parser_tools.h
+++ b/include/vfs/Tools/vfs_parser_tools.h
@@ -37,10 +37,10 @@ namespace vfs
 	{
 		static const vfs::size_t BUFFER_SIZE = 1024;
 	public:
-		CReadLine(vfs::tReadableFile& rFile);
+		CReadLine(vfs::tReadableFile& rFile, bool autoControlFile = true);
 		~CReadLine();
 
-		bool fillBuffer();
+		bool fillBuffer(bool refill = false);
 		bool fromBuffer(std::string& line);
 		bool getLine(std::string& line);
 	private:
@@ -49,7 +49,7 @@ namespace vfs
 		vfs::size_t			_bytes_left;
 		vfs::size_t			_buffer_pos;
 		vfs::size_t			_buffer_last;
-		bool				_eof;
+		bool				_auto_ctrl_file;  // if false the object user checks for EOF and closes file
 
 		void operator=(CReadLine const& rl);
 	};
diff --git a/src/Core/vfs_string.cpp b/src/Core/vfs_string.cpp
index 3f1fea84f..0b8940925 100644
--- a/src/Core/vfs_string.cpp
+++ b/src/Core/vfs_string.cpp
@@ -162,7 +162,7 @@ void vfs::String::as_utf16(std::string const& str, vfs::String::str_t &str16)
 	}
 	catch(utf8::invalid_utf8& ex)
 	{
-		utf8::uint8_t c = ex.utf8_octet();
+		utf8::utfchar8_t c = ex.utf8_octet();
 		VFS_THROW( _BS(L"Invalid UTF8 character '") << (wchar_t)c << L"'=" << (unsigned char)c <<_BS::wget );
 	}
 	catch(utf8::not_enough_room &ex)
@@ -185,12 +185,17 @@ vfs::String::str_t vfs::String::as_utf16(const char* str)
 }
 void vfs::String::as_utf16(const char* str, vfs::String::str_t &str16)
 {
+	wchar_t buf16[1024];
+
 	if(str == NULL)
 	{
 		return;
 	}
 	try
 	{
+		memset(buf16, 0, 1024 * sizeof(wchar_t));
+		mbstowcs(buf16, str, 1024 - 1);
+
 		::size_t len = strlen(str);
 		::size_t d = utf8::distance(str,str+len);
 		if(d > 0)
@@ -202,7 +207,8 @@ void vfs::String::as_utf16(const char* str, vfs::String::str_t &str16)
 	catch(utf8::invalid_utf8& ex)
 	{
-		utf8::uint8_t c = ex.utf8_octet();
+		utf8::utfchar8_t c = ex.utf8_octet();
-		VFS_THROW( _BS(L"Invalid UTF8 character '") << (wchar_t)c << L"'=" << (unsigned char)c << _BS::wget );
+		//VFS_THROW( _BS(L"Invalid UTF8 character '") << (wchar_t)c << L"'=" << (unsigned char)c << _BS::wget );
+		VFS_THROW(_BS(L"Invalid UTF8 character '") << (wchar_t)c << L"'=" << (unsigned char)c << L" str16:" << buf16 << _BS::wget);
 	}
 	catch(utf8::not_enough_room &ex)
 	{
diff --git a/src/Tools/vfs_parser_tools.cpp b/src/Tools/vfs_parser_tools.cpp
index dbd7c74bc..99c8e8204 100644
--- a/src/Tools/vfs_parser_tools.cpp
+++ b/src/Tools/vfs_parser_tools.cpp
@@ -34,36 +34,42 @@
 /*************************************************************************************/
 /*************************************************************************************/
 
-vfs::CReadLine::CReadLine(vfs::tReadableFile& rFile)
-: _file(rFile), _buffer_pos(0), _eof(false)
+vfs::CReadLine::CReadLine(vfs::tReadableFile& rFile, bool autoControlFile)
+	: _file(rFile), _buffer_pos(0), _auto_ctrl_file(autoControlFile)
 {
 	memset(_buffer,0,sizeof(_buffer));
 
 	vfs::COpenReadFile rfile(&_file);
-	_bytes_left = rfile->getSize();
+	vfs::size_t startReadPosition = _file.getReadPosition();
+	_bytes_left = startReadPosition >= rfile->getSize() ? 0 : rfile->getSize() - startReadPosition;
 	fillBuffer();
-	vfs::UByte utf8bom[3] = {0xef,0xbb,0xbf};
-	if(memcmp(utf8bom, &_buffer[0],3) == 0)
+
+	if (startReadPosition == 0)
 	{
-		_buffer_pos += 3;
+		vfs::UByte utf8bom[3] = { 0xef, 0xbb, 0xbf };
+		if (memcmp( utf8bom, &_buffer[0], 3) == 0 )
+		{
+			_buffer_pos += 3;
+		}
 	}
 	rfile.release();
 };
 
 vfs::CReadLine::~CReadLine()
 {
-	if(_file.isOpenRead())
+	if (_auto_ctrl_file && _file.isOpenRead())
 	{
 		_file.close();
 	}
 }
 
-bool vfs::CReadLine::fillBuffer()
+bool vfs::CReadLine::fillBuffer(bool refill)
 {
-	if(_eof)
+	if (_bytes_left == 0 || (_auto_ctrl_file == false && refill == true))
 	{
 		return false;
 	}
+
 	vfs::size_t bytesRead = BUFFER_SIZE < _bytes_left ? BUFFER_SIZE : _bytes_left;
 	try
 	{
@@ -79,7 +85,6 @@ bool vfs::CReadLine::fillBuffer()
 	}
 
 	_bytes_left -= bytesRead;
-	_eof = (_bytes_left == 0);
 
 
 	// bite-wise read files usually terminate a line with \n (or \r\n on WIN32)
@@ -136,7 +141,7 @@ bool vfs::CReadLine::fromBuffer(std::string& line)
 					}
 					else
 					{
-						done = !fillBuffer();
+						done = !fillBuffer(true);
 					}
 				}
 				else if(*temp == '\n' || *temp == 0)
@@ -148,12 +153,12 @@ bool vfs::CReadLine::fromBuffer(std::string& line)
 			}
 			else
 			{
-				done = !fillBuffer();
+				done = !fillBuffer(true);
 			}
 		}
 		else
 		{
-			done = !fillBuffer();
+			done = !fillBuffer(true);
 		}
 	}
 	return false;
@@ -162,7 +167,16 @@ bool vfs::CReadLine::fromBuffer(std::string& line)
 bool vfs::CReadLine::getLine(std::string& line)
 {
 	line.clear();
-	return fromBuffer(line);
+	bool gotLine = fromBuffer(line);
+	
+	// If file handler is controlled by caller, we have to read to EOL or EOF. So if happened to read over EOL (due to buffering),
+	// then we need to move caret back to a position where just fetched line ends (EOL).
+	if (_auto_ctrl_file == false && _buffer_pos < _buffer_last)
+	{
+		vfs::offset_t offset = (vfs::offset_t)_buffer_pos - _buffer_last;  // "lesser - greater" intentionally to get negative offset
+		_file.setReadPosition(offset, _file.SD_CURRENT);
+	}
+	return gotLine;
 }
 
 /*************************************************************************************/
